### DISCUSSION TITLE
First contact: got Daffodil connected to VS Code debugger via DAP

### DISCUSSION THREAD

**Discussion by arosien** [jw3/example-daffodil-debug#10] [Show and tell]
I've got a super-basic "step debug a parse using VS Code via DAP" working (see attached animated gif). It relays the current schema--no "stack" yet--and data position. Stepping works, but I haven't implemented "continue" + "pause" yet.

![sampleWorkspace-1619731568700](https://user-images.githubusercontent.com/31163/117070595-cdc8b000-ace2-11eb-952f-6d8fcba33961.gif)

**Comment by arosien** [comment]
I've added more daffodil state to the "Variables" pane and now correctly handle looking at the state of the "schema stack" trace.

![sampleWorkspace-1620419866815](https://user-images.githubusercontent.com/31163/117506091-e1218880-af39-11eb-9c5f-f71d13645436.gif)

**Comment by arosien** [comment]
Also pushed a working Pause/Continue in #12. If you have a small data file you need to press it quick!

**Comment by arosien** [comment]
I've got setting and updating breakpoints working:

![sampleWorkspace-1621898706449](https://user-images.githubusercontent.com/31163/119418446-dd696200-bcac-11eb-9eb1-d1867cfdf45a.gif)

**Comment by arosien** [comment]
Trying current data region as DAP "output event", whacha think?

![sampleWorkspace-1622003573433](https://user-images.githubusercontent.com/31163/119602896-03702e80-bda1-11eb-920c-6d0a4c2164da.gif)

**Reply by jw3** [reply]
> whacha think?

:exploding_head: wow! that looks good.

Trying to be critical I would say people have also mentioned binary representation of the data, and I am really disappointed it didn't make this cut :upside_down_face:

**Reply by arosien** [reply]
Oh, I didn't commit this! I can if that works for folks.

**Reply by jw3** [reply]
I think it provides a great entrypoint to talking about the custom views that are needed. "We have this data on the backend and can send it over, need to build displays for it..."  There are of course other implementation details we have to work through related to how it is sent, but showing the potential now is nice.

On the frontend a view that sets its text rather than appending might be a simple first custom view impl.  Tying that view into an event that delivers this text would make for a good first take on this.

**Reply by arosien** [reply]
I'm happy that it didn't take much code to add this feature. Roughly:

```diff
+++ b/src/main/scala/ddb/debugger/dapodil/DAPodil.scala
@@ -353,6 +353,7 @@ object DAPodil extends IOApp {
   trait Debugee {
     def data(): Signal[IO, Data]
     def state(): Stream[IO, Debugee.State]
+    def outputs(): Stream[IO, Events.OutputEvent]
 
     def step(): IO[Unit]
     def continue(): IO[Unit]
```

**Reply by jw3** [reply]
Nicely done.

> There are of course other implementation details we have to work through related to how it is sent

Was thinking like `OutputEvent` vs `DaffodilInputStreamPositionEvent` or variable, or whatever is best way to provide specific event for input stream position.

**Reply by arosien** [reply]
BTW the data view is just from some code I ripped out of the "legacy" `InteractiveDebugger`, I didn't make a hex view on my own :).

**Comment by arosien** [comment]
FYI, it's super-simple to replicate the `InteractiveDebugger` command-line experience, it's just an `evaluate` request/response in DAP. I jammed an infoset display in about 10 minutes:

![sampleWorkspace-1622055286697](https://user-images.githubusercontent.com/31163/119715844-4e2d8d00-be19-11eb-85d9-484f26c6b53b.gif)
